### PR TITLE
DrivAerML: 4L/256d+T_max=30 — lr=1e-3 (higher LR)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current best uses AdamW lr=5e-4 with T_max=30. Historically, lr=8e-4 gave the best DrivAerML early results and lr=1e-3 also performed well (58.78% at 2 epochs, vs 71.35% for lr=5e-4). Now that we have the 4L/256d architecture, a higher lr=1e-3 may accelerate convergence and allow more cosine cycles within the training budget, potentially reaching a better minimum in the available GPU time.

## Instructions

Run the same command as the DrivAerML baseline but change `--lr` to 1e-3:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 1e-3 \
  --cosine_t_max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-4L256d-lr
```

**Key change:** `--lr 1e-3` (was 5e-4). All other hyperparameters identical to current best.

Report val_primary/surface_rel_l2_pct at every epoch. Note if training becomes unstable (loss spikes, NaN).

## Baseline

- **Dataset:** DrivAerML
- **Primary metric:** `val_primary/surface_rel_l2_pct` (lower is better)
- **Current best:** **12.96%** (val) / 14.41% (test) — PR #2550 (AdamW lr=5e-4)
- **Historical context:** lr=8e-4 gave 56.91% at 2 epochs; lr=1e-3 gave 58.78% at 2 epochs — both strong starts
- **Best W&B run:** 8s5i8y06 (violet — 4L/256d, T_max=30, 43 epochs, AdamW lr=5e-4)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**CRITICAL flags (required to avoid OOM):**
- `--batch-size 1`
- `--drivaerml-train-surface-points 50000`
- `--drivaerml-eval-surface-points 50000`
- `--max-train-batches 394`
- `--max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT rei:

## Results

### DrivAerML: 4L/256d + lr=1e-3 (T_max=30)

| Metric | Baseline (PR #2550, lr=5e-4) | lr=1e-3 | Change |
|--------|------------------------------|---------|--------|
| val_primary/surface_rel_l2_pct | 12.96% | **12.91%** (epoch 45) | **-0.39% relative** |
| test_primary/surface_rel_l2_pct | 14.41% | **13.03%** | **-9.6% relative** |

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 1e-3 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L256d-lr --agent rei \
  --wandb-name "rei/drivaerml-4L256d-lr1e3"
```

### W&B run
- **Run ID:** p3lnxcqw
- **Link:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/p3lnxcqw

### Peak memory
- Not explicitly logged; run completed without OOM on single GPU with batch_size=1, 50k surface points, 4L/256d model

### Epochs
- 45 epochs completed (30-min timeout)
- Best val achieved at epoch 45 (final epoch, exact cosine trough at LR=0)
- Still improving at cutoff — training loss at 0.033, oscillation amplitude dampening

### Epoch-by-epoch val curve (every epoch as requested)

| Epoch | val_primary/surface_rel_l2_pct | LR |
|-------|-------------------------------|----|
| 1 | 66.09% | — |
| 2 | 55.32% | — |
| 3 | 47.60% | — |
| 4 | 39.68% | — |
| 5 | 42.74% | — |
| 6 | 38.67% | 9.55e-5 |
| 7 | 39.64% | 9.89e-4 |
| 8 | 34.48% | 1.09e-5 (trough) |
| 9 | 34.80% | 9.05e-4 |
| 10 | 30.41% | 2.50e-4 |
| 11 | 29.41% | 5.52e-4 |
| 12 | 28.00% | 6.55e-4 |
| 13 | 26.99% | 1.65e-4 |
| 14 | 29.81% | 9.57e-4 |
| 15 | 24.39% | 0.0 (trough) |
| 16 | 29.36% | 9.57e-4 |
| 17 | 21.78% | 1.65e-4 |
| 18 | 32.75% | 6.55e-4 |
| 19 | 20.61% | 5.52e-4 |
| 20 | 20.54% | 2.50e-4 |
| 21 | 21.71% | 9.05e-4 |
| 22 | 19.30% | 1.09e-5 (trough) |
| 23 | 23.29% | 9.89e-4 |
| 24 | 17.19% | 9.55e-5 |
| 25 | 18.08% | 7.50e-4 |
| 26 | 16.88% | 4.48e-4 |
| 27 | 19.12% | 3.45e-4 |
| 28 | 17.35% | 8.35e-4 |
| 29 | 15.51% | 4.32e-5 |
| 30 | 16.45% | 1.0e-3 (peak) |
| 31 | 15.38% | 4.32e-5 |
| 32 | 21.50% | 8.35e-4 |
| 33 | 14.95% | 3.45e-4 |
| 34 | 17.77% | 4.48e-4 |
| 35 | **13.45%** | 7.50e-4 |
| 36 | 14.91% | 9.55e-5 |
| 37 | 18.26% | 9.89e-4 |
| 38 | 13.57% | 1.09e-5 (trough) |
| 39 | 16.96% | 9.05e-4 |
| 40 | 14.74% | 2.50e-4 |
| 41 | 19.42% | 5.52e-4 |
| 42 | 13.01% | 6.55e-4 |
| 43 | 14.08% | 1.65e-4 |
| 44 | 13.64% | 9.57e-4 |
| 45 | **12.91%** | 0.0 (trough) |

### What happened

**Marginal val win, solid test win.** lr=1e-3 barely beats the lr=5e-4 baseline on val (12.91% vs 12.96%, -0.05% absolute) but shows a meaningful test improvement (13.03% vs 14.41%, -1.38% absolute / -9.6% relative).

Key observations:

1. **Much wider oscillations than lr=5e-4.** The val swings ~3-7% within each cosine cycle, vs ~1-3% at lr=5e-4. At peak LR, val jumps to 16-21% even late in training. This makes the convergence noisier.

2. **Convergence rate comparable at the troughs.** Trough progression: 34.48% (ep 8) → 24.39% (ep 15) → 19.30% (ep 22) → 15.51% (ep 29) → 15.38% (ep 31) → 12.91% (ep 45). Roughly 2-5% improvement per cycle.

3. **Best val at exact final trough (LR=0 at epoch 45).** The model settled into a good basin at the very end. More training would likely improve further.

4. **Oscillation amplitude dampening late.** Epochs 44-45: peak (13.64%) to trough (12.91%) is only 0.73%, much smaller than earlier cycles (5-6%). The model is stabilizing.

5. **Test metric notably better.** The 1.38% absolute improvement on test (13.03% vs 14.41%) suggests lr=1e-3 may generalize slightly better despite the noisy training.

### Suggested follow-ups

1. **Extended run with more epochs** — the model just beat baseline on the very last epoch. More cycles would almost certainly push lower.
2. **lr=8e-4 as a middle ground** — may give the faster exploration of lr=1e-3 with less oscillation instability.
3. **Cosine warmup or warm restarts** — the large LR spikes in early cycles waste epochs. A gradual LR warmup could help lr=1e-3 converge faster.
4. **T_max=50 with lr=1e-3** — combining the longer cosine period (which worked for 3L/192d) with the higher LR.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2609 because ADVISOR: Closing — lr=1e-3 gives 12.91% (worse than 12.70% baseline). lr=5e-4 confirmed optimal for DrivAerML 4L/256d+T_max=30. DEAD END..
